src/Nethermind/Nethermind.Blockchain/BlockhashProvider.cs

```diff
@@ -46,26 +46,7 @@ public BlockhashProvider(IBlockFinder blockTree, ISpecProvider specProvider, IWo
                 return null;
             }
 
-            BlockHeader header = _blockTree.FindParentHeader(currentBlock, BlockTreeLookupOptions.TotalDifficultyNotNeeded) ??
-                throw new InvalidDataException("Parent header cannot be found when executing BLOCKHASH operation");
-
-            for (var i = 0; i < _maxDepth; i++)
-            {
-                if (number == header.Number)
-                {
-                    if (_logger.IsTrace) _logger.Trace($"BLOCKHASH opcode returning {header.Number},{header.Hash} for {currentBlock.Number} -> {number}");
-                    return header.Hash;
-                }
-
-                header = _blockTree.FindParentHeader(header, BlockTreeLookupOptions.TotalDifficultyNotNeeded);
-                if (header is null)
-                {
-                    throw new InvalidDataException("Parent header cannot be found when executing BLOCKHASH operation");
-                }
-            }
-
-            if (_logger.IsTrace) _logger.Trace($"BLOCKHASH opcode returning null for {currentBlock.Number} -> {number}");
-            return null;
+            return _blockTree.FindBlockHash(number);
```

Comment:
There was a bug before where the canonical block was not correct. Hence, why it iterate one by one.

Reply:
Changed the caching to only update number when on mainchain (so block inserts alone don't update number cache)

Changed test to more explicitly check around this; also included the pivot forked block in when updating mainchain in test (check this if correct @asdacap); as it was updating mainchain post the changing block but not including the changing block as part of branch update

Reply:
You mean if no UpdateMainChain?

Reply:
Fixed the legacy blockhashes by adding it back as switch; but this test is cursed

Reply:
@asdacap better?

Reply:
You think there is a situation where the main chain is being changed while validating a mainchain block in the evm?